[WIP/RFT] ramips: add support for HC5761A, improve support for HC5661A and HC5861B

dengqf6 opened the pull request:
For hardware specs, HiWiFi HC5761A is very similar to HC5761, the only difference is HC5761A uses MT7628AN SoC instead
For flash instructions please refer to HC5661A or HC5861B

- Create HC5X61A.dtsi for all HiWiFi MT7628AN models (HC5661A, HC5761A, HC5861B)
- Explicitly disable EHCI and OHCI nodes of HC5661A and HC5861B
- Correct image size of HC5661A (15808k)
- Fix pinctrl of HC5661A
- Use GPIO interrupt keys instead of polling
- Use tpt trigger for wireless LEDs
- Increase SPI frequency to 80MHz

Known bug:
- SD slot doesn't work (#1500)

981213 commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -71,12 +71,21 @@ define Device/glinet_vixmini
   DTS := HC5661A
-  IMAGE_SIZE := $(ralink_default_fw_size_16M)
+  IMAGE_SIZE := 15808k
+  SUPPORTED_DEVICES := hc5661a
```
In theory this will break sysupgrade on hc5661a. (I don't have time to verify my assumption though.)
sysupgrade checks board name against SUPPORTED_DEVICES and since you've changed board name to hiwifi,hc5661a this match will fail.
It should be SUPPORTED_DEVICES += hc5661a instead.

981213 commented:
And I think you should split this commit into multiple ones. Each changes described in your commit message deserves a separated commit, and adding HC5761A support should be another commit with required commit message for adding a new device support.

981213 commented:
Ah...
Actually I mean splitting commits this way:

Create HC5X61A.dtsi **and** use it for HC5661A and HC5861B
Fix stuff for HC5661A/HC5861B
Add support for HC5761A

And the commit adding HC5761A should still have a commit message describing this device and flash instruction (Even if it's just copy&paste from HC5661A/HC5861B)

adrianschmutzler commented:
Your HC5761A patch seems to miss the Makefile changes.
Note that part of the reorganizations are also planned in https://github.com/openwrt/openwrt/pull/2171

adrianschmutzler commented:
There has been a reorganization in ramips target today. Please:
- update your board.d subfiles since they have been re-sorted
- update your Makefile (mt****/rt****.mk) changes so sorting of devices is correct
- rename your DTS/DTSI files following the scheme soc_vendor_device.dts(i), e.g. mt7620a_tplink_archer-c50-v1.dts (and update includes there if necessary)
- remove `DTS := ...` from Makefile
- add `MTK_SOC` to Makefile (needs to match "soc" in DTS file name), e.g. `MTK_SOC := mt7620a`

Thank you!

Note that for your case that will obsolete some of the renames and actually might make patches easier. Be sure to update DTSI names and includes.

981213 commented on target/linux/ramips/dts/mt7628an_hiwifi_hc5661a.dts:
```diff
@@ -1,127 +1,46 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
```
Oops...
To specify a license for an existing dts, you'd need to ask for an Acked-by from every one who contributed to the original file :(
I think you can drop this line for now and use a separated patch for this task later.

981213 replied:
@dengqf6 Would you mind if I merge this PR with this line dropped?

981213 commented on target/linux/ramips/dts/mt7628an_hiwifi_hc5861b.dts:
```diff
@@ -1,46 +1,24 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
```
Same as above.

981213 commented:
dropped SPDX License Identifier for HC5661A/HC5861B, tested on HiWiFi HC5661A and merged into my staging tree at: https://git.openwrt.org/?p=openwrt/staging/981213.git

I'll push it tomorrow if nobody disapprove it :)

adrianschmutzler commented on target/linux/ramips/base-files/etc/board.d/01_leds:
```diff
@@ -132,7 +131,8 @@ d-team,newifi-d2)
-gehua,ghl-r-001)
+gehua,ghl-r-001|\
+hiwifi,hc5861b)
 	ucidef_set_led_netdev "internet" "internet" "$boardname:blue:internet" "eth0.2"
```
I cannot find "$boardname:blue:internet" being defined for hiwifi,hc5861b anywhere. Did I overlook it or is this a mistake?

981213 replied:
Should be a mistake. It's hc5661a which needs to be added here. fixed that in my staging tree...

981213 commented:
Dropped it from my staging tree. My modified version can be found here: https://git.openwrt.org/?p=openwrt/staging/981213.git;a=shortlog;h=refs/heads/pr2155

dengqf6 commented:
Yes that's why I marked this pull request as RFT

adrianschmutzler commented:
Which of the netif devices matches the sticker/label/case/box MAC address?

Despite, I wonder about the MAC address setup anyway:
```
lan_mac=`mtd_get_mac_ascii bdinfo "Vfac_mac "`
[ -n "$lan_mac" ] || lan_mac=$(cat /sys/class/net/eth0/address)
```

First, one should improve this to get rid of the ticks:
`lan_mac=$(mtd_get_mac_ascii bdinfo "Vfac_mac ")`

Second, why this fallback:
`[ -n "$lan_mac" ] || lan_mac=$(cat /sys/class/net/eth0/address)`

This only makes sense if the first line sometimes returns an empty string (so this line still provides the base address for wan_mac.

Commit:
ramips: add HC5X61A.dtsi for HiWiFi MT7628AN boards

HiWiFi has several MT7628AN routers which have similar specs
Add HC5X61A.dtsi to include them, like HC5X61.dtsi (for MT7620A)

Signed-off-by: DENG Qingfang <dengqf6@mail2.sysu.edu.cn>

Commit:
ramips: improve support for HiWiFi HC5661A and HC5861B

HC5661A:
- Fix pinctrl
- Fix image size (15808k)
- Use switch trigger for WAN LED

Both:
- Use tpt LED trigger for wireless
- Explicitly disable USB nodes

Signed-off-by: DENG Qingfang <dengqf6@mail2.sysu.edu.cn>

Commit:
ramips: add support for HiWiFi HC5761A

HiWiFi HC5761A is an "MT7628AN variant" of HC5761

Specifications:
- MediaTek MT7628AN 580MHz
- 128 MB DDR2 RAM
- 16 MB SPI Flash
- 2.4G MT7628AN 802.11bgn 2T2R 300Mbps
- 5G MT7610EN 802.11ac 433Mbps
- 3x 10/100 Mbps Ethernet

Flash instruction:
1. Get SSH access to the router
2. SSH to router with `ssh -p 1022 root@192.168.199.1`, The SSH password is the same as the webconfig one
3. Upload OpenWrt sysupgrade firmware into the router's `/tmp` folder with SCP
4. Run `mtd write /tmp/<filename> firmware`
5. reboot

Known bug:
- SD slot does not work (See PR 1500)

Signed-off-by: DENG Qingfang <dengqf6@mail2.sysu.edu.cn>

dengqf6 commented:
@adrianschmutzler The label MAC is at `Vfac_mac `
@981213 You may push your SD driver fix for HC5661A first, then I'll include it here for HC5761A.

adrianschmutzler commented:
@dengqf6 

> The label MAC is at Vfac_mac

Thanks. For HC5761A only, or for all three devices mentioned?

Any idea about why the fallback is necessary?

`[ -n "$lan_mac" ] || lan_mac=$(cat /sys/class/net/eth0/address)`

dengqf6 commented:
For all HiWiFi boards

981213 commented:
> You may push your SD driver fix for HC5661A first, then I'll include it here for HC5761A.

That one is a hack and I don't really want to push it now :)

981213 commented:
Since there's no tester for HC5861B appearing for half a month, could you reduce SPI clock for HC5861B down to 32MHz (This is what's used in MTK OpenWrt SDK for mt76x8) so that we could have this PR merged?

dengqf6 commented:
@981213 HC5861B uses the same flash chip as HC5661A's (w25q128) so I think it will work

981213 commented:
I was thinking that a poorly designed PCB may cause problems when bus speed is high enough but I have no idea exactly how fast when PCB trace length matching is necessary to be done. Now after some actual searching I found out that SPI is actually considered "low speed" :P

Merged into my staging tree (again...) at https://git.openwrt.org/?p=openwrt/staging/981213.git . Thanks!
